File: python/packUpdate/services/version_service.py
"""Version service for updating project version"""
import json
import os
import re
from typing import Optional
from ..utils.logger import log

class VersionService:
    @staticmethod
    def update_project_version(project_path: str, version_type: str, quiet_mode: bool = False) -> None:
        """Update project version in package.json and package-lock.json"""
        try:
            package_json_path = os.path.join(project_path, 'package.json')
            
            if not os.path.exists(package_json_path):
                log('⚠️  No package.json found, skipping version update')
                return

            with open(package_json_path, 'r') as f:
                package_json = json.load(f)
            
            current_version = package_json.get('version', '1.0.0')
            
            if VersionService._is_valid_semver(version_type):
                new_version = version_type
            else:
                new_version = VersionService._calculate_new_version(current_version, version_type)
                if not new_version:
                    log(f'❌ Invalid version type: {version_type}. Use major, minor, patch, or x.y.z format')
                    return

            # Update package.json
            package_json['version'] = new_version
            with open(package_json_path, 'w') as f:
                json.dump(package_json, f, indent=2)
                f.write('\n')
            
            # Update package-lock.json if it exists
            package_lock_path = os.path.join(project_path, 'package-lock.json')
            if os.path.exists(package_lock_path):
                with open(package_lock_path, 'r') as f:
                    package_lock = json.load(f)
                
                package_lock['version'] = new_version
                if 'packages' in package_lock and '' in package_lock['packages']:
                    package_lock['packages']['']['version'] = new_version
                
                with open(package_lock_path, 'w') as f:
                    json.dump(package_lock, f, indent=2)
                    f.write('\n')

            log(f'📦 Updated project version: {current_version} → {new_version}')
            
        except Exception as error:
            log(f'❌ Failed to update project version: {error}')

    @staticmethod
    def _is_valid_semver(version: str) -> bool:
        """Check if string is valid semver format (x.y.z)"""
        semver_regex = r'^\d+\.\d+\.\d+$'
        return bool(re.match(semver_regex, version))

    @staticmethod
    def _calculate_new_version(current_version: str, version_type: str) -> Optional[str]:
        """Calculate new version based on current version and type"""
        try:
            parts = [int(x) for x in current_version.split('.')]
            if len(parts) != 3:
                return None

            major, minor, patch = parts

            if version_type.lower() == 'major':
                major += 1
                minor = 0
                patch = 0
            elif version_type.lower() == 'minor':
                minor += 1
                patch = 0
            elif version_type.lower() == 'patch':
                patch += 1
            else:
                return None

            return f'{major}.{minor}.{patch}'
        except (ValueError, IndexError):
            return None

File: node/src/services/versionService.ts
import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';
import { log } from '../utils/logger';

export class VersionService {
  /**
   * Update project version in package.json and package-lock.json
   */
  static updateProjectVersion(projectPath: string, versionType: string, quietMode: boolean = false): void {
    try {
      const packageJsonPath = path.join(projectPath, 'package.json');
      
      if (!fs.existsSync(packageJsonPath)) {
        log('⚠️  No package.json found, skipping version update');
        return;
      }

      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
      const currentVersion = packageJson.version || '1.0.0';
      
      let newVersion: string;
      
      if (this.isValidSemver(versionType)) {
        // Specific version provided
        newVersion = versionType;
      } else {
        // Calculate new version based on type
        const calculatedVersion = this.calculateNewVersion(currentVersion, versionType);
        if (!calculatedVersion) {
          log(`❌ Invalid version type: ${versionType}. Use major, minor, patch, or x.y.z format`);
          return;
        }
        newVersion = calculatedVersion;
      }

      // Update package.json
      packageJson.version = newVersion;
      fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2) + '\n');
      
      // Update package-lock.json if it exists
      const packageLockPath = path.join(projectPath, 'package-lock.json');
      if (fs.existsSync(packageLockPath)) {
        const packageLock = JSON.parse(fs.readFileSync(packageLockPath, 'utf8'));
        packageLock.version = newVersion;
        if (packageLock.packages && packageLock.packages[""]) {
          packageLock.packages[""].version = newVersion;
        }
        fs.writeFileSync(packageLockPath, JSON.stringify(packageLock, null, 2) + '\n');
      }

      log(`📦 Updated project version: ${currentVersion} → ${newVersion}`);
      
    } catch (error) {
      log(`❌ Failed to update project version: ${error}`);
    }
  }

  /**
   * Check if string is valid semver format (x.y.z)
   */
  private static isValidSemver(version: string): boolean {
    const semverRegex = /^\d+\.\d+\.\d+$/;
    return semverRegex.test(version);
  }

  /**
   * Calculate new version based on current version and type
   */
  private static calculateNewVersion(currentVersion: string, type: string): string | null {
    const parts = currentVersion.split('.').map(Number);
    if (parts.length !== 3 || parts.some(isNaN)) {
      return null;
    }

    let [major, minor, patch] = parts;

    switch (type.toLowerCase()) {
      case 'major':
        major++;
        minor = 0;
        patch = 0;
        break;
      case 'minor':
        minor++;
        patch = 0;
        break;
      case 'patch':
        patch++;
        break;
      default:
        return null;
    }

    return `${major}.${minor}.${patch}`;
  }
}
